### ISSUE TITLE
Persistent Syncer connection failures not exposed for readiness reporting

### ISSUE THREAD

**Issue by fasaxc** [projectcalico/libcalico-go#478]
If the KDD Syncer's watch ends with a persistent network error then it just keeps retrying (with warning logs) but it doesn't tell its client that it's having trouble.  Ideally, consistent failures would be exposed in some way, so that we can feed them into the healthcheck and report Felix/Typha as not ready.

## Possible Solution
- Could have the Syncer optionally register with the healthcheck
- Could have the Syncer spot consistent failures and either resync or flag it using the status API.

## Steps to Reproduce (for bugs)

1. Let syncer come into sync
2. kill the API server
3. Syncer logs warnings but keeps retrying
4.

## Context
This was desired behaviour for the Felix/Typha readiness feature

## TODO

- [ ] resolve in libcalico-go
- [ ] add k8sfv tests

**Comment by caseydavenport** [comment]
@fasaxc does this affect KDD and etcd or just KDD?

Could this just be triggered by the Syncer indicating "WaitForDatastore"?

**Comment by fasaxc** [comment]
Yeah, that sounds like a good option.

**Comment by caseydavenport** [comment]
Went to have a look a reproducing this today and saw something perhaps interesting (or maybe I'm just tired?)

I noticed that following the repro steps of killing the apiserver _did _ result in Felix going to not-ready in KDD mode. Looking closer though, it only seems to do so after receiving an update over the syncer that says the IPIP tunnel address has been deleted, triggering Felix to restart.

```
2019-03-08 03:10:11.737 [INFO][42] resource.go 260: Target config /etc/calico/confd/config/bird_ipam.cfg has been updated
2019-03-08 03:10:11.730 [INFO][3095] watchercache.go 242: Failed to create watcher ListRoot="/calico/resources/v3/projectcalico.org/nodes" error=Get https://10.96.0.1:443/api/v1/nodes?resourceVersion=5344&watch=true: dial tcp 10.96.0.1:443: connect: connection refused performFullResync=false
2019-03-08 03:10:11.730 [INFO][3095] watchercache.go 242: Failed to create watcher ListRoot="/calico/resources/v3/projectcalico.org/networkpolicies" error=Get https://10.96.0.1:443/apis/networking.k8s.io/v1/networkpolicies?resourceVersion=4985&watch=true: dial tcp 10.96.0.1:443: connect: connection refused performFullResync=false
2019-03-08 03:10:11.731 [INFO][3095] watchercache.go 242: Failed to create watcher ListRoot="/calico/resources/v3/projectcalico.org/clusterinformations" error=Get https://10.96.0.1:443/apis/crd.projectcalico.org/v1/clusterinformations?resourceVersion=4985&watch=true: dial tcp 10.96.0.1:443: connect: connection refused performFullResync=false
2019-03-08 03:10:11.732 [INFO][3095] watchercache.go 242: Failed to create watcher ListRoot="/calico/resources/v3/projectcalico.org/globalnetworksets" error=Get https://10.96.0.1:443/apis/crd.projectcalico.org/v1/globalnetworksets?resourceVersion=4985&watch=true: dial tcp 10.96.0.1:443: connect: connection refused performFullResync=false
2019-03-08 03:10:11.733 [INFO][3095] watchercache.go 242: Failed to create watcher ListRoot="/calico/resources/v3/projectcalico.org/globalnetworkpolicies" error=Get https://10.96.0.1:443/apis/crd.projectcalico.org/v1/globalnetworkpolicies?resourceVersion=4985&watch=true: dial tcp 10.96.0.1:443: connect: connection refused performFullResync=false
2019-03-08 03:10:11.733 [INFO][3095] watchercache.go 242: Failed to create watcher ListRoot="/calico/resources/v3/projectcalico.org/felixconfigurations" error=Get https://10.96.0.1:443/apis/crd.projectcalico.org/v1/felixconfigurations?resourceVersion=4985&watch=true: dial tcp 10.96.0.1:443: connect: connection refused performFullResync=false
2019-03-08 03:10:11.733 [INFO][3095] customresource.go 282: Error listing resources error=Get https://10.96.0.1:443/apis/crd.projectcalico.org/v1/hostendpoints: dial tcp 10.96.0.1:443: connect: connection refused
2019-03-08 03:10:11.733 [INFO][3095] watchercache.go 188: Failed to perform list of current data during resync ListRoot="/calico/resources/v3/projectcalico.org/hostendpoints" error=Get https://10.96.0.1:443/apis/crd.projectcalico.org/v1/hostendpoints: dial tcp 10.96.0.1:443: connect: connection refused
2019-03-08 03:10:11.734 [INFO][3095] config_batcher.go 61: Host config update for this host: {{HostConfig(node=casey-crc-kadm-ms,name=IpInIpTunnelAddr) <nil>  0s} 3}
2019-03-08 03:10:11.734 [INFO][3095] config_batcher.go 112: Sending config update global: map[LogSeverityScreen:Info IpInIpEnabled:true ReportingIntervalSecs:0 ClusterGUID:69c74524a0264bfd8dceeb9a5ee668e2 ClusterType:k8s,bgp,kdd CalicoVersion:v3.5.2], host: map[].
2019-03-08 03:10:11.734 [INFO][3095] event_sequencer.go 210: Possible config update. global=map[string]string{"CalicoVersion":"v3.5.2", "LogSeverityScreen":"Info", "IpInIpEnabled":"true", "ReportingIntervalSecs":"0", "ClusterGUID":"69c74524a0264bfd8dceeb9a5ee668e2", "ClusterType":"k8s,bgp,kdd"} host=map[string]string{}
2019-03-08 03:10:11.734 [INFO][3095] config_params.go 220: Merging in config from datastore (global): map[CalicoVersion:v3.5.2 LogSeverityScreen:Info IpInIpEnabled:true ReportingIntervalSecs:0 ClusterGUID:69c74524a0264bfd8dceeb9a5ee668e2 ClusterType:k8s,bgp,kdd]
```

I'm a bit confused where this is coming from - based on the logs it's sorta out of the blue. I will retry with debug and see if anything looks more clear.

**Comment by caseydavenport** [comment]
Ah! Figured it out. With debugging on I spotted that the watchersyncer was sending a bunch of deleted through for the node objects, seemingly it thought that the nodes had disappeared!

Tracing it back, turns out we weren't returning an error on a failed List() call for nodes only...I think this will do the trick: https://github.com/projectcalico/libcalico-go/pull/1057

I'll try with that, expecting that persistent syncer errors will no longer result in Felix reporting unready...
